drivers/gpio/gpio_shell.c

```diff
@@ -1,187 +1,449 @@
 /*
  * Copyright (c) 2018 Intel Corporation
  * Copyright (c) 2021 Dennis Ruffer <daruffer@gmail.com>
+ * Copyright (c) 2023 Nick Ward <nix.ward@gmail.com>
  *
  * SPDX-License-Identifier: Apache-2.0
- *
- * Use "device list" command for GPIO port names
  */
 
-#include <zephyr/sys/printk.h>
+#include <zephyr/drivers/gpio.h>
 #include <zephyr/shell/shell.h>
-#include <zephyr/init.h>
-#include <string.h>
+
 #include <stdio.h>
-#include <zephyr/drivers/gpio.h>
-#include <stdlib.h>
-#include <ctype.h>
-#include <zephyr/logging/log.h>
 
-#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+#define ARGV_DEV 1
+#define ARGV_PIN 2
+#define ARGV_CONF 3
+#define ARGV_VALUE 3
 
-LOG_MODULE_REGISTER(gpio_shell);
+#define NGPIOS_UNKNOWN -1
+#define PIN_NOT_FOUND UINT8_MAX
 
-struct args_index {
-	uint8_t port;
-	uint8_t index;
-	uint8_t mode;
-	uint8_t value;
+/* Pin syntax maximum length */
+#define PIN_SYNTAX_MAX 32
+
+struct gpio_ctrl {
+	const struct device *dev;
+	int8_t ngpios;
+	gpio_port_pins_t reserved_mask;
+	const char **line_names;
+	uint8_t line_names_len;
+	const union shell_cmd_entry *subcmd;
 };
 
-static const struct args_index args_indx = {
-	.port = 1,
-	.index = 2,
-	.mode = 3,
-	.value = 3,
+struct sh_gpio {
+	const struct device *dev;
+	gpio_pin_t pin;
 };
 
-static int cmd_gpio_conf(const struct shell *sh, size_t argc, char **argv)
+/*
+ * Find idx-th pin reference from the set of non reserved
+ * pin numbers and provided line names.
+ */
+static void port_pin_get(gpio_port_pins_t reserved_mask, const char **line_names,
+			 uint8_t line_names_len, size_t idx, struct shell_static_entry *entry)
```

Comment:
Yes true, I was thinking at the time it would be good to give the user the option pin number or line name but at the same time during testing I stopped using the pin numbers and was using line names exclusively. Happy to use line name exclusively when available and fall back to pin numbers when line name is not available unless anyone else disagrees.